### PR TITLE
Bbox subsetting for GLAH06, GLAH14, ATL03, ATL08 and GEDI02_A

### PR DESCRIPTION
**Pull request by alex-s-gardner**
GLAH06 + new tests

### CODE REVIEW

**Comment by alex-s-gardner** [comment]
@evetion this PR should be good to merged

File: src/ICESat-2/ATL03.jl [hunk 1]
```diff
@@ -72,61 +84,99 @@ function points(
...
+        segment = segment,
+        track = Fill(track, length(datetime)),
+        strong_beam = Fill(atlas_beam_type == "strong", length(datetime)),
+        sun_angle = sun_angle,
+        detector_id = Fill(parse(Int8, spot_number), length(datetime)),
+        height_reference = height_ref,
     )
-end
-
-function map_counts(values, counts)
```
Comment (evetion) [inline src/ICESat-2/ATL03.jl:120]: Was this removal on purpose? This function is quite useful to map ATL08 classifications to ATL03.

File: src/ICESat-2/ATL03.jl [hunk 2]
```diff
@@ -182,3 +232,20 @@ function create_mapping(dfsegment, unique_segments)
     end
     index_map
 end
+
+# fast map between segment and photon
+function segment2photon(seg_cnt)
```
Comment (evetion) [inline src/ICESat-2/ATL03.jl:237]: This replaces `map_counts`? But no ATL08 code is in this PR.

**Comment by alex-s-gardner** [comment]
This was just a bit faster than create_mapping so I updated it.

Sent from my iPhone

On Nov 23, 2022, at 11:11 PM, Maarten Pronk ***@***.***> wrote:

﻿

@evetion commented on this pull request.

________________________________

In src/ICESat-2/ATL03.jl<https://urldefense.us/v3/__https://github.com/evetion/SpaceLiDAR.jl/pull/33*discussion_r1031122466__;Iw!!PvBDto6Hs4WbVuu7!cCDtZrz2u6bLP-BiIuV8eEvITexgPMKTD5kJiNeEWmIir7U7U7u-Nw6iSZJsW_jmdE169547$>:

> @@ -182,3 +232,20 @@ function create_mapping(dfsegment, unique_segments)
     end
     index_map
 end
+
+# fast map between segment and photon
+function segment2photon(seg_cnt)


This replaces map_counts? But no ATL08 code is in this PR.

**Comment by evetion** [comment]
It's great that you go through all the code and place comments, I will make those into issues later on.

I have to give a closer look at the change in `map_counts`, great if it's faster (also because it's non-repeating), but ideally I add a test in master so we're sure the results are the same (and those of classify as well).

**Comment by evetion** [comment]
> @evetion can you review this. I'd like to add more PR's but it would be good not to drift too far from Master.

Sorry for the delay, bit busy with reports. We can have a dev branch, or multiple branches based of one another, so you can carry on, although good git tooling really helps with that. Or are you unsure that I don't accept these changes? In that case, we can discuss your plans, so we're on the same page.

**Review by evetion — DISMISSED** [review]

**Comment by alex-s-gardner** [comment]
> Sorry for the delay, bit busy with reports. We can have a dev branch, or multiple branches based of one another, so you can carry on, although good git tooling really helps with that. Or are you unsure that I don't accept these changes? In that case, we can discuss your plans, so we're on the same page.

My git skills are pretty rudimentary, though I'm hoping they will improve. My main concern is that I get too far ahead of what you've looked at and approved as my pull requests build on one another. I also don't want my pull request to become to large that they are difficult to review.

As for best practice should I:
1. fork `A` from `main`
2. make some manageable amount of updates
3. issue a request to have `A` pulled into `main`
4. fork `B` from `A` and continue on ?

File: src/ICESat-2/ATL03.jl [hunk 3]
```diff
@@ -72,61 +84,99 @@ function points(
...
-    segment = read(file, "$track/geolocation/segment_id")::Vector{Int32}
-    segments = map_counts(segment, segment_counts)[1:step:end]
+    # extract data posted at segment frequency and map to photon frequency
+    segment = file["$track/geolocation/segment_id"][:]::Vector{Int32}
+    segment = segment[ph_ind]
 
-    sun_angle = read(file, "$track/geolocation/solar_elevation")::Vector{Float32}
-    sun_angles = map_counts(sun_angle, segment_counts)[1:step:end]
+    sun_angle = file["$track/geolocation/solar_elevation"][:]::Vector{Float32}
+    sun_angle = sun_angle[ph_ind]
```
[the remark this answers is not in the document]
Comment (evetion) [reply src/ICESat-2/ATL03.jl:149]: It doesn't work indeed.

**Review by evetion — DISMISSED** [review]
I do really dislike the amount of code it now takes to get some data, but overall it's a good improvement, thanks!

**Comment by evetion** [comment]
> My git skills are pretty rudimentary, though I'm hoping they will improve. My main concern is that I get too far ahead of what you've looked at and approved as my pull requests build on one another. I also don't want my pull request to become to large that they are difficult to review.

This review was hard to review in an online setting, due to the many lines and linelength, sorry it took me so long.

> As for best practice should I:
> 
>     1. fork `A` from `main`
> 
>     2. make some manageable amount of updates
> 
>     3. issue a request to have `A` pulled into `main`
> 
>     4. fork `B` from `A` and continue on ?

That could work, with rebasing your work from B on master once A is merged. It helps if the work in B is different than the work done in A.

---

For now, I've added the bbox subsetting for ATL08 and improved the tests a bit in [`1464e5c` (#33)](https://github.com/evetion/SpaceLiDAR.jl/pull/33/commits/1464e5c45c8fc336c5fe37357ffd05408325de20). I've also merged master, so if you're happy with this, we can merge this PR.

Happy new year 🎉

**Comment by alex-s-gardner** [comment]
This all looks good to me. This is a big update... things should be much more stable hereafter.

As a note, two things I've noticed:
[1] dispatch for each datatype creates a huge amount of duplicate code.. previously I wrote similar code in Matlab using a single function with numerous conditional statements to handle differences between dataset structure.

[2] I'm not a big fan of changing the package name using  `const` e.g. `const SL = SpaceLiDAR`... my preference is explicit over concise. I'm hyper conscious of this as NASA is awash in acronyms that makes it difficult for outsiders to understand our jargon.

**Comment by evetion** [comment]
> [1] dispatch for each datatype creates a huge amount of duplicate code.. previously I wrote similar code in Matlab using a single function with numerous conditional statements to handle differences between dataset structure. No action needed... just an observation

I agree with the amount of code, but do you have a specific function/piece of code in mind? I was thinking (no concrete plan) that we ideally have a generic hdf5 points reader and we only specify the fieldnames (height="place/in/hdf5") we want, generic code solving catting everything together. That would also solve users wanting more/less fields than we now hardcode, although applying filtering correctly/directly could be problematic.

> [2] I'm not a big fan of changing the package name using `const` e.g. `const SL = SpaceLiDAR`... my preference is explicit over concise. I'm hyper conscious of this as NASA is awash in acronyms that makes it difficult for outsiders to understand our jargon.

Fair point, but this change is only for the tests, it's not in the documentation, so it's up to the user?

**Comment by alex-s-gardner** [comment]
> was thinking (no concrete plan) that we ideally have a generic hdf5 points reader and we only specify the fieldnames (height="place/in/hdf5") we want, generic code solving catting everything together. That would also solve users wanting more/less fields than we now hardcode, although applying filtering correctly/directly could be problematic.

Ya, I think having a function like `h5point("path/to/file.h5", ["var1", "var2", "var3"])` that returns a nested named tuple with the same structure as the original h5 would be a good start. But that will be a big lift to implement in SpaceLiDAR
